vllm/v1/worker/gpu_model_runner.py

```diff
@@ -1017,20 +1019,56 @@ def execute_model(
         scheduler_output: "SchedulerOutput",
         intermediate_tensors: Optional[IntermediateTensors] = None,
     ) -> Union[ModelRunnerOutput, torch.Tensor]:
-        # Update KVConnector with the KVConnector metadata forward().
-        if has_kv_transfer_group():
-            get_kv_transfer_group().bind_connector_metadata(
-                scheduler_output.kv_connector_metadata)
+
+        def maybe_setup_kv_connector():
+            # Update KVConnector with the KVConnector metadata forward().
+            if has_kv_transfer_group():
+                kv_connector = get_kv_transfer_group()
+                assert isinstance(kv_connector, KVConnectorBase_V1)
+                assert scheduler_output.kv_connector_metadata is not None
+                kv_connector.bind_connector_metadata(
+                    scheduler_output.kv_connector_metadata)
+
+                # Background KV cache transfers happen here.
+                # These transfers are designed to be async and the requests
+                # involved may be disjoint from the running requests.
+                # Do this here to save a collective_rpc.
+                kv_connector.start_load_kv(get_forward_context())
+
+        def maybe_wait_for_save():
+            if has_kv_transfer_group():
+                kv_connector = get_kv_transfer_group()
+                kv_connector.wait_for_save()
+
+        def maybe_get_finished() -> tuple[set[str], set[str]]:
+            if has_kv_transfer_group():
+                kv_connector = get_kv_transfer_group()
+                return kv_connector.get_finished()
+            else:
+                return set(), set()
 
         self._update_states(scheduler_output)
         if not scheduler_output.total_num_scheduled_tokens:
+            # KV send/recv even if no work to do.
+            with set_forward_context(None, self.vllm_config):
```

Comment:
Why do we need to set the forward context here?

Reply:
The forward context gets passed into the KV connector.

We need to change this as it's not a stable API. And we should do this very soon before there's more KVConnector adoption